src/plugins/data_view_field_editor/public/components/field_editor/composite_editor.tsx

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React from 'react';
+import { EuiNotificationBadge, EuiFlexGroup, EuiFlexItem, EuiSpacer, EuiText } from '@elastic/eui';
+import { ScriptField } from './form_fields';
+import { useFieldEditorContext } from '../field_editor_context';
+import { useFieldPreviewContext } from '../preview';
+import { UseArray, UseField, TextField } from '../../shared_imports';
+import { TypeField } from './form_fields';
+import { RUNTIME_FIELD_OPTIONS_PRIMITIVE } from './constants';
+
+export const CompositeEditor = ({}) => {
+  const { links, existingConcreteFields } = useFieldEditorContext();
+  const { fields } = useFieldPreviewContext();
+  // const { fields } = useFieldPreviewContext();
+  // const names = fields.map((field) => field.key);
+  // console.log('fields from context', fields);
+  // what I want to do -
+  // 1. get list of fields
+  // 2. get types from `fields` in context
+
+  return (
+    <div>
+      <ScriptField existingConcreteFields={existingConcreteFields} links={links} />
+      <EuiSpacer size="xl" />
+      <UseArray path="fields">
+        {({ form }) => {
+          console.log('Render subfield list', fields);
```

Comment:
You are not using the `fields` provided by `<UseArray />`. If not then you probably don't need UseArray.

UseArray is meant to handle adding and removing fields dynamically. If you know in advance the fields then you can simply loop over them and create `<UseField />` (as you seem to be doing here).

This is what you should have as children of `<UseArray />`

https://github.com/elastic/kibana/blob/main/src/plugins/es_ui_shared/static/forms/hook_form_lib/components/__stories__/use_array_basic.tsx#L49

Reply:
With that said I _do_ think you need `<UseArray >`. Those fields should be declared in the form.

So you don't want to access the fields like this

```
const { fields } = useFieldPreviewContext();
```

But you want to have

```jsx
<UseArray path="compositeFields" />
  {(items) => {
    // loop over the items and render `<UseField />`
  }}
<UseArray />
```

When you load an existing runtime field you probably will have in your form `defaultValue` (or should be using a `deserializer` to have those fields available)

```
const { form } = useForm({
  defaultValue: {
    title: 'foo,.
   ...
   compositeFields: [
     { name: 'foo', type: 'number' },
     { name: 'bar', type: 'boolean' },
     ...
   ]
  }
});
```

So what happens after fetching the preview from the server? You will need to update the form value

```
form.updateFieldValues({ compositeFields: [<new-array>] });
```

https://github.com/elastic/kibana/blob/main/src/plugins/es_ui_shared/static/forms/hook_form_lib/components/__stories__/use_array_dynamic_data.tsx#L276

Reply:
Something I just discovered that I was getting tripped up on for quite a while - `setFieldValue` needs a different key when an array is used - `fields` becomes `fields__array__`. Once I started using `setFieldValue('fields__array__', value);` then I started to see progress.

Reply:
I wouldn't mess with the `fields__array__`. That's implementation details and could change in the future. The public api is `updateFieldValues()`